src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -702,13 +697,56 @@ impl RustcTargetData {
             target_config.insert(ct, host_config.clone());
         }
 
-        Ok(RustcTargetData {
+        let mut res = RustcTargetData {
             rustc,
+            config,
+            requested_kinds: requested_kinds.into(),
             host_config,
             host_info,
             target_config,
             target_info,
-        })
+        };
+
+        // Get all kinds we currently know about.
+        //
+        // For now, targets can only ever come from the root workspace
+        // units as artifact dependencies are not a thing yet, so this
+        // correctly represents all the kinds that can happen. When we
+        // have artifact dependencies or other ways for targets to
+        // appear at places that are not the root units, we may have
+        // to revisit this.
+        let all_kinds = requested_kinds
+            .iter()
+            .copied()
+            .chain(ws.members().flat_map(|p| {
+                p.manifest()
+                    .default_kind()
+                    .into_iter()
+                    .chain(p.manifest().forced_kind())
+            }));
+        for kind in all_kinds {
+            if let CompileKind::Target(target) = kind {
+                match res.target_config.entry(target) {
+                    std::collections::hash_map::Entry::Occupied(_) => (),
+                    std::collections::hash_map::Entry::Vacant(place) => {
```

Comment:
I think it's ok to use `contains_key` followed by `insert` here to avoid the verbose-ness of the `Entry`

Reply:
Right, these maps will probably stay small so the overhead should be minimal :)